docs/canonical_testing.md

```diff
@@ -0,0 +1,186 @@
+k8stest verify  # Run the verify tests
+k8stest e2e-gce  # Run the e2e tests on gce
+k8stest e2e-gke-serial  # Run the serial tests on gke
+```
+
+Customizing how a specific job runs:
+```
+k8stest test-go -- pkg/api pkg/kubelet # Run only the api and kubelet unit tests
+k8stest test-integration -- Deadline  # Run only Deadline integration tests
+k8stest verify -- godeps  # Run only the godep verification
+k8stest e2e-gce -- --push --test  # Push to an existing cluster and run tests
+k8stest e2e-gke-serial -- --skip=Daemon  # Skip the Daemon tests
+k8stest e2e-gce-example -- --focus=rollover  # Only run the rollover tests
+```
+
+Customizing where things run:
+```
+k8stest --mode=docker  # Run tests inside a local container
+k8stest --mode=k8s --config=provider-config.yaml # Schedule a pod to run tests
+k8stest --mode=local --upload=upload-config.yaml # Upload results somewhere
+```
+
+
+## job class commands
+
+Right now we have a bunch of useful scripts like `hack/e2e.go`,
+`hack/verify-all.sh` and the like.
+
+We will make these the official entry point that defines what happens for each
+job class. This also defines the public interface between what users can call
+and the raw `kubectl`, `go test`, etc commands issued during the job.
+
+In other words we will expect developers (and automation) to either a) call
+these commands directly or else make the raw calls that these scripts are
+composed of.
+
+Specifically accessing any intermediate scripts like `ginkgo-e2e.sh` will no
+longer be supported and may atrophe over time without warning.
+
+Many jobs typically follow a similar pattern. For example essentially every
+`kubernetes-e2e-*` job issues the same `hack/e2e.go --build --up --test --down`
+command with different extra arguments and environment variables. Therefore
+these are encapsulated in the e2e class.
+
+
+## job images
+
+Each class of job will have a companion image that:
+
+* contains all the dependencies required to complete the job class
+* logic to deal with customization that varies per job:
+  - job class customization lives in a well-known folder
+* logic to deal with customization per user:
+  - provider customization (project, creds, etc) lives in another well-known folder
+  - upload customization (same but for uploads) lives in a third well-known folder
+
+For example the e2e images will contain everything to run `hack/e2e.go` as well
+as whatever else remains the same between different jobs in the class.
+
+Each image will expect the customization to live in a particular location. The
+image will
+
+* process this customized information
+* set the customized arguments/environment variables
+* run the job class command included in the image.
+
+## Yaml customization
+
+The way each job will customize the class is through a yaml file. The yaml file
+will specify things that the image needs to customize -- arguments and
+environment variables.
+
+When jobs run locally the top-level runner will duplicate the logic that
```

Comment:
The default will be local as it is faster and more flexible for developers.

Real runs will happen on kubernetes.

The point/goal is to make the interface between users/docker and tests trivial. Right now we have lots of logic spread around everywhere so it is hard to figure out exactly what needs to happen to run tests.

Otherwise there really isn't much point to the docker mode as it will never actually be used, but it may be useful as a way for a developer to have a slightly more confidence that this will work in the official location on kubernetes without waiting to launch a kubernetes cluster.